hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/RatisReplicationCheckHandler.java

```diff
@@ -0,0 +1,198 @@
+    // than remainingRedundancy which ECContainerReplicaCount uses.
+    RatisContainerReplicaCount replicaCount =
+        new RatisContainerReplicaCount(container, replicas, pendingAdd,
+            pendingDelete, requiredNodes, minReplicasForMaintenance);
+
+    ContainerPlacementStatus placementStatus =
+        getPlacementStatus(replicas, requiredNodes, Collections.EMPTY_LIST);
+
+    ContainerPlacementStatus placementStatusWithPending = placementStatus;
+    if (replicaPendingOps.size() > 0) {
+      placementStatusWithPending =
+          getPlacementStatus(replicas, requiredNodes, replicaPendingOps);
+    }
+    boolean sufficientlyReplicated
+        = replicaCount.isSufficientlyReplicated(false);
+    boolean isPolicySatisfied = placementStatus.isPolicySatisfied();
+    if (!sufficientlyReplicated || !isPolicySatisfied) {
+      ContainerHealthResult.UnderReplicatedHealthResult result =
+          new ContainerHealthResult.UnderReplicatedHealthResult(
+          container, replicaCount.getRemainingRedundancy(),
+          isPolicySatisfied && replicas.size() >= requiredNodes,
+          replicaCount.isSufficientlyReplicated(true),
+          replicaCount.isUnrecoverable());
+      result.setMisReplicated(!isPolicySatisfied)
+          .setMisReplicatedAfterPending(
+              !placementStatusWithPending.isPolicySatisfied())
+          .setDueToMisReplication(
+              !isPolicySatisfied && replicaCount.isSufficientlyReplicated());
+      return result;
+    }
+
+    boolean isOverReplicated = replicaCount.isOverReplicated(false);
+    if (isOverReplicated) {
+      boolean repOkWithPending = !replicaCount.isOverReplicated(true);
+      return new ContainerHealthResult.OverReplicatedHealthResult(
+          container, replicaCount.getExcessRedundancy(false), repOkWithPending);
+    }
+    // No issues detected, just return healthy.
+    return new ContainerHealthResult.HealthyResult(container);
+  }
+
+  /**
+   * Given a set of ContainerReplica, transform it to a list of DatanodeDetails
+   * and then check if the list meets the container placement policy.
+   * @param replicas List of containerReplica
+   * @param replicationFactor Expected Replication Factor of the containe
+   * @return ContainerPlacementStatus indicating if the policy is met or not
+   */
+  private ContainerPlacementStatus getPlacementStatus(
+      Set<ContainerReplica> replicas, int replicationFactor,
+      List<ContainerReplicaOp> pendingOps) {
+
+    Set<DatanodeDetails> replicaDns = replicas.stream()
+        .map(ContainerReplica::getDatanodeDetails)
+        .collect(Collectors.toSet());
+    for (ContainerReplicaOp op : pendingOps) {
+      if (op.getOpType() == ContainerReplicaOp.PendingOpType.ADD) {
+        replicaDns.add(op.getTarget());
+      } else if (op.getOpType() == ContainerReplicaOp.PendingOpType.DELETE) {
+        replicaDns.remove(op.getTarget());
+      }
```

Comment:
Can there be a case where there is pending ADD & pending DELETE to the same node? Some kind of race condition.

Reply:
Better to use a Map<DataNodeDetails,Integer> in that case.

Reply:
This should not be able to happen, as if a node has a replica it cannot get another copy of it. For a delete to be scheduled it must have a copy which will prevent an add etc.

However I will change this to two IF statements rather than if .. else if